test/infer/test_infer_util.py

```diff
@@ -96,6 +96,35 @@ def guide(data):
     assert_allclose(jnp.mean(obs_pred), 0.8, atol=0.05)
 
 
+def test_predictive_with_particles():
+    num_particles = 5
+    num_samples = 2
+    fdim = 3
+    num_data = 10
+
+    def model(x, y=None):
+        latent = numpyro.sample("latent", dist.Normal(0.0, jnp.ones(fdim)).to_event(1))
+        with numpyro.plate("data", x.shape[0]):
+            numpyro.sample("y", dist.Normal(jnp.matmul(x, latent), 1.0), obs=y)
+
+    def guide(x, y=None):
+        latent_loc = numpyro.param(
+            "latent_loc", jnp.ones(fdim), constraint=constraints.real
+        )
+        numpyro.sample("latent", dist.Normal(latent_loc, 1.0).to_event(1))
```

Comment:
Maybe we can add an assertion here that `latent_loc.ndim == 1` (those assertations are typical in numpyro models). We need not to restrict users to write "batch-compatible" code. Is it a special requirement of SteinVI? Either way, we might not need to restrict ourselves in this more user-faced utility.

Reply:
I can add an assertion.

From the context I'm not sure what you mean by batch-compatible. Would you care to elaborate?

Reply:
I see. Thanks for clarifying. You are right; I'll have to vmap to produce the desired behavior. SteinVI vmaps over the particle dimension(s), so that shouldn't be a problem. I'll add a shape assert to one of the unittest to confirm in #833.

Reply:
Maybe we can rename the current `__call__` method into `_call_with_params(self, params)`. And in the `__call__` method, we can check
```
# NOTE: we also need to handler `rng_key`, `args`, `kwargs` correctly.
if self.guide is None or (batch_ndims == 0):
    return self._call_with_params(self.params)
elif batch_ndims == 1:
    tree_map(lambda x: jnp.swap_axes(x, 0, 1), vmap(self._call_with_params)(self.params))
else:
    raise NotImplementedError
```

Reply:
I was thinking to handle it with a similar conditional in the `__call__` instead. If we use your suggestions I can factor the `sample_shape` out the new `__call__`.

Could we so 
``` python            
jax.vmap(self._call_with_params, in_axes=self.batch_ndims, out_axes=-self.batch_ndims)(self.params)
```
instead of the `tree_map`?

Reply:
Yeah, using `in_axes` and `out_axes` are much cleaner. Maybe `in_axes=0` and `out_axes=1` for `batch_ndims==1`? What's your definition of `batch_ndims`? Currently, `batch_ndims` is used to denote the number of batch dimensions of `posterior_samples`. Here we don't have `posterior_samples` but having `guide`. My intuition is to let it denote the number of batch dimensions of `params`, but would be interested in hearing your intuition.

Reply:
I think of `batch_ndims` as the dimensionality of the particles, which is equivalent to what you describe. 

I sketched an implementation based on your suggestion. I have to compute  `param_batch_shape` to find the number of `rng_key’s needed. Not sure I can avoid this; what do you think? Your suggestion means we can avoid tuple `num_samples`. I’ll update to use `in_axes` and `out_axes` tomorrow morning.